File: python/galileo/utils/api_client.py
from importlib.metadata import version
from os import getenv
from time import time
from typing import Any, Optional

import jwt
import logging

from galileo_core.constants.request_method import RequestMethod
from galileo_core.helpers.execution import async_run
from galileo_core.helpers.api_client import ApiClient as CoreApiClient
from galileo.constants.routes import Routes
from galileo.schema.trace import TracesIngestRequest, TracesIngestResponse
from galileo.utils.request import HttpHeaders, make_request
from httpx import Client, AsyncClient, HTTPError, Response, Timeout
from urllib.parse import urljoin

_logger = logging.getLogger(__name__)


class ApiClient:
    project_id: Optional[str] = None
    log_stream_id: Optional[str] = None

    def __init__(self, project_name: str, log_stream_name: Optional[str] = None):
        self.project_id = None
        self.api_url = self.get_api_url()
        if self.healthcheck():
            self.token = self.get_token()
            project = self.get_project_by_name(project_name)
            if project is None:
                self.project_id = self.create_project(project_name)["id"]
                _logger.info(
                    f"🚀 Creating new project... project {project_name} created!"
                )
            else:
                if project["type"] != "gen_ai":
                    raise Exception(
                        f"Project {project_name} is not a Galileo 2.0 project"
                    )
                self.project_id = project["id"]

            log_stream = self.get_log_stream_by_name(
                project_id=self.project_id, log_stream_name=log_stream_name
            )
            if log_stream is None:
                self.log_stream_id = self.create_log_stream(
                    project_id=self.project_id, log_stream_name=log_stream_name
                )["id"]
                _logger.info(
                    f"🚀 Creating new log stream... log stream {log_stream_name} created!"
                )
            else:
                self.log_stream_id = log_stream["id"]

    def get_api_url(self) -> str:
        console_url = getenv("GALILEO_CONSOLE_URL")
        if console_url is None:
            # TODO: Set to the multi-tenant cluster when it's ready
            raise Exception("GALILEO_CONSOLE_URL must be set")
        if any(map(console_url.__contains__, ["localhost", "127.0.0.1"])):
            api_url = "http://localhost:8088"
        else:
            api_url = console_url.replace("console", "api")
        return api_url

    def get_token(self) -> str:
        api_key = getenv("GALILEO_API_KEY")
        if api_key:
            return self.api_key_login(api_key).get("access_token", "")

        raise Exception("GALILEO_API_KEY must be set")

    def healthcheck(self) -> bool:
        async_run(
            make_request(
                RequestMethod.GET, base_url=self.base_url, endpoint=Routes.healthcheck
            )
        )
        return True

    def api_key_login(self, api_key: str) -> dict[str, str]:
        return async_run(
            make_request(
                RequestMethod.POST,
                base_url=self.base_url,
                endpoint=Routes.api_key_login,
                json={"api_key": api_key},
            )
        )

    @property
    def base_url(self) -> str:
        return self.api_url

    @property
    def auth_header(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self.token}"}

    def _make_request(
        self,
        request_method: RequestMethod,
        endpoint: str,
        json: Optional[dict] = None,
        data: Optional[dict] = None,
        files: Optional[dict] = None,
        params: Optional[dict] = None,
        json_request_only: bool = False,
    ) -> Any:
        # Check to see if our token is expired before making a request
        # and refresh token if it's expired
        if endpoint not in [Routes.login, Routes.api_key_login] and self.token:
            claims = jwt.decode(self.token, options={"verify_signature": False})
            if claims.get("exp", 0) < time():
                self.token = self.get_token()

        if json_request_only:
            content_headers = HttpHeaders.accept_json()
        else:
            content_headers = HttpHeaders.json()
        headers = {**self.auth_header, **content_headers}
        return ApiClient.make_request_sync(
            request_method=request_method,
            base_url=self.base_url,
            endpoint=endpoint,
            json=json,
            data=data,
            files=files,
            params=params,
            headers=headers,
        )

    async def _make_async_request(
        self,
        request_method: RequestMethod,
        endpoint: str,
        json: Optional[dict] = None,
        data: Optional[dict] = None,
        files: Optional[dict] = None,
        params: Optional[dict] = None,
        json_request_only: bool = False,
    ) -> Any:
        # Check to see if our token is expired before making a request
        # and refresh token if it's expired
        if endpoint not in [Routes.login, Routes.api_key_login] and self.token:
            claims = jwt.decode(self.token, options={"verify_signature": False})
            if claims.get("exp", 0) < time():
                self.token = self.get_token()

        if json_request_only:
            content_headers = HttpHeaders.accept_json()
        else:
            content_headers = HttpHeaders.json()
        headers = {**self.auth_header, **content_headers}
        await make_request(
            request_method=request_method,
            base_url=self.base_url,
            endpoint=endpoint,
            json=json,
            data=data,
            files=files,
            params=params,
            headers=headers,
        )

    async def ingest_traces(
        self, traces_ingest_request: TracesIngestRequest
    ) -> dict[str, str]:
        traces_ingest_request.log_stream_id = self.log_stream_id
        json = traces_ingest_request.model_dump()

        return await self._make_async_request(
            RequestMethod.POST,
            endpoint=Routes.traces.format(project_id=self.project_id),
            json=json,
        )

    def ingest_traces_sync(
        self, traces_ingest_request: TracesIngestRequest
    ) -> dict[str, str]:
        traces_ingest_request.log_stream_id = self.log_stream_id
        json = traces_ingest_request.model_dump()

        return self._make_request(
            RequestMethod.POST,
            endpoint=Routes.traces.format(project_id=self.project_id),
            json=json,
        )

    def get_project_by_name(self, project_name: str) -> Any | None:
        projects = self._make_request(
            RequestMethod.GET,
            endpoint=Routes.projects,
            params={"project_name": project_name},
        )
        if len(projects) < 1:
            return None
        return projects[0]

    def create_project(self, project_name: str) -> dict[str, str]:
        return self._make_request(
            RequestMethod.POST,
            endpoint=Routes.projects,
            # Default project type is gen_ai (Galileo 2.0)
            json={"name": project_name, "type": "gen_ai"},
        )

    def get_log_stream_by_name(
        self, project_id: str, log_stream_name: str
    ) -> Any | None:
        log_streams = self._make_request(
            RequestMethod.GET,
            # TODO: update to an endpoint that allows filtering by name
            endpoint=Routes.log_streams.format(project_id=project_id),
        )
        if len(log_streams) < 1:
            return None
        for log_stream in log_streams:
            if log_stream["name"] == log_stream_name:
                return log_stream
        return None

    def create_log_stream(
        self, project_id: str, log_stream_name: str
    ) -> dict[str, str]:
        return self._make_request(
            RequestMethod.POST,
            endpoint=Routes.log_streams.format(project_id=project_id),
            json={"name": log_stream_name},
        )

    # TODO: Move this method to galileo_core.helpers.api_client
    @staticmethod
    def make_request_sync(
        request_method: RequestMethod,
        base_url: str,
        endpoint: str,
        skip_ssl_validation: bool = False,
        read_timeout: float = 60.0,
        **kwargs: Any,
    ) -> Any:
        url = urljoin(base_url, endpoint)
        with Client(
            base_url=base_url,
            timeout=Timeout(read_timeout, connect=5.0),
            verify=not skip_ssl_validation,
        ) as client:
            response = client.request(
                method=request_method.value, url=url, timeout=read_timeout, **kwargs
            )
            CoreApiClient.validate_response(response)
            return response.json()

File: js/src/api-client.ts
import { decode } from "jsonwebtoken";

import axios, { AxiosRequestConfig, AxiosResponse, Method } from "axios";
import { Project, ProjectTypes } from "./types/project.types.js";
import { Routes } from "./types/routes.types";

import querystring from "querystring";
import createClient, { Client } from "openapi-fetch";
import type { components, paths } from "./types/api.types.js";

export enum RequestMethod {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
}
import { promises as fs } from "fs";

type DatasetFormat = components["schemas"]["DatasetFormat"];
export type ListDatasetResponse = components["schemas"]["ListDatasetResponse"];
export type DatasetContent = components["schemas"]["DatasetContent"];
export type Dataset = components["schemas"]["DatasetDB"];
export type DatasetRow = components["schemas"]["DatasetRow"];

type CollectionPaths =
  | paths["/datasets"]
  | paths["/datasets/{dataset_id}/content"];
type CollectionResponse = ListDatasetResponse | DatasetContent;

export class GalileoApiClient {
  public type: ProjectTypes | undefined = undefined;
  public projectId: string = "";
  public runId: string = "";
  public datasetId: string = "";
  private apiUrl: string = "";
  private token: string = "";
  private client: Client<paths> | undefined = undefined;

  public async init(
    projectName?: string | undefined,
    datasetId?: string,
  ): Promise<void> {
    this.apiUrl = this.getApiUrl();
    if (await this.healthCheck()) {
      this.token = await this.getToken();
      this.client = createClient({
        baseUrl: this.apiUrl,
        headers: { Authorization: `Bearer ${this.token}` },
      });

      if (datasetId) {
        this.datasetId = datasetId;
      }

      if (projectName) {
        try {
          this.projectId = await this.getProjectIdByName(projectName);
          // eslint-disable-next-line no-console
          console.log(`✅ Using ${projectName}`);
        } catch (err: unknown) {
          const error = err as Error;

          if (error.message.includes("not found")) {
            const project = await this.createProject(projectName);
            this.projectId = project.id;
            // eslint-disable-next-line no-console
            console.log(`✨ ${projectName} created.`);
          } else {
            throw err;
          }
        }
      }
    }
  }

  private processResponse<T>(data: T | undefined, error: object | unknown): T {
    if (data) {
      return data;
    }

    if (error) {
      if (typeof error === "object" && "detail" in error) {
        throw new Error(`Request failed: ${JSON.stringify(error.detail)}`);
      }

      throw new Error(`Request failed: ${JSON.stringify(error)}`);
    }

    throw new Error("Request failed");
  }

  private async fetchAllPaginatedItems<
    Path extends CollectionPaths,
    Response extends CollectionResponse,
    Item,
  >(
    path: "/datasets" | "/datasets/{dataset_id}/content",
    extractItems: (response: Response) => Item[],
    params: Path["get"]["parameters"],
  ): Promise<Item[]> {
    if (!this.client) {
      throw new Error("Client not initialized");
    }

    let items: Item[] = [];
    let startingToken: number | null = 0;

    do {
      const updatedParams: Record<string, unknown> = {
        path: params.path,
        query: { ...params.query, starting_token: startingToken },
      };

      const { data, error } = await this.client.GET(path, {
        params: updatedParams,
      });

      const collection = this.processResponse(data as Response, error);
      items = items.concat(extractItems(collection));
      startingToken = collection.next_starting_token ?? null;
    } while (startingToken !== null);

    return items;
  }

  private getApiUrl(): string {
    const consoleUrl = process.env.GALILEO_CONSOLE_URL;

    if (!consoleUrl) {
      throw new Error("❗ GALILEO_CONSOLE_URL must be set");
    }

    if (consoleUrl.includes("localhost") || consoleUrl.includes("127.0.0.1")) {
      return "http://localhost:8088";
    } else {
      return consoleUrl.replace("console", "api");
    }
  }

  private async healthCheck(): Promise<boolean> {
    return await this.makeRequest<boolean>(
      RequestMethod.GET,
      Routes.healthCheck,
    );
  }

  private async getToken(): Promise<string> {
    const apiKey = process.env.GALILEO_API_KEY;

    if (apiKey) {
      const loginResponse = await this.apiKeyLogin(apiKey);
      return loginResponse.access_token || "";
    }

    const username = process.env.GALILEO_USERNAME;
    const password = process.env.GALILEO_PASSWORD;

    if (username && password) {
      const loginResponse = await this.usernameLogin(username, password);
      return loginResponse.access_token || "";
    }

    throw new Error(
      "❗ GALILEO_API_KEY or GALILEO_USERNAME and GALILEO_PASSWORD must be set",
    );
  }

  private async apiKeyLogin(
    api_key: string,
  ): Promise<{ access_token: string }> {
    return await this.makeRequest<{ access_token: string }>(
      RequestMethod.POST,
      Routes.apiKeyLogin,
      {
        api_key,
      },
    );
  }

  private async usernameLogin(username: string, password: string) {
    return await this.makeRequest<{ access_token: string }>(
      RequestMethod.POST,
      Routes.login,
      querystring.stringify({
        username,
        password,
      }),
    );
  }

  private async getProjectIdByName(project_name: string): Promise<string> {
    const projects = await this.makeRequest<Project[]>(
      RequestMethod.GET,
      Routes.projects,
      null,
      {
        project_name,
        type: this.type,
      },
    );

    if (projects.length < 1) {
      throw new Error(`Galileo project ${project_name} not found`);
    }

    return projects[0].id;
  }

  private async createProject(project_name: string): Promise<{ id: string }> {
    return await this.makeRequest<{ id: string }>(
      RequestMethod.POST,
      Routes.projects,
      {
        name: project_name,
        type: this.type,
      },
    );
  }

  public getDatasets = async (): Promise<Dataset[]> => {
    return await this.fetchAllPaginatedItems<
      paths["/datasets"],
      ListDatasetResponse,
      Dataset
    >(
      "/datasets",
      (response: ListDatasetResponse) => response.datasets ?? [],
      {},
    );
  };

  public async createDataset(
    name: string,
    filePath: string,
    format: DatasetFormat,
  ): Promise<Dataset> {
    if (!this.client) {
      throw new Error("Client not initialized");
    }

    const fileBuffer: Buffer = await fs.readFile(filePath);
    const blob: Blob = new Blob([fileBuffer]);
    const formdata = new FormData();
    formdata.append("file", blob, name);

    const { data, error } = await this.client.POST("/datasets", {
      params: { query: { format } },
      // @ts-expect-error openapi-typescript does not properly translate FormData for uploading files - https://github.com/openapi-ts/openapi-typescript/issues/1214
      body: formdata,
      bodySerializer: (body) => {
        // define a custom serializer to prevent openapi-fetch from serializing the FormData object as JSON
        return body;
      },
    });

    const dataset = this.processResponse(data, error) as Dataset;
    // eslint-disable-next-line no-console
    console.log(
      `✅  Dataset '${dataset.name}' with ${dataset.num_rows} rows uploaded.`,
    );
    return dataset;
  }

  public async getDatasetContent(datasetId: string): Promise<DatasetRow[]> {
    if (!this.client) {
      throw new Error("Client not initialized");
    }

    return await this.fetchAllPaginatedItems<
      paths["/datasets/{dataset_id}/content"],
      DatasetContent,
      DatasetRow
    >(
      `/datasets/{dataset_id}/content`,
      (response: DatasetContent) => response.rows ?? [],
      { path: { dataset_id: datasetId } },
    );
  }

  private getAuthHeader(token: string): { Authorization: string } {
    return { Authorization: `Bearer ${token}` };
  }

  private validateResponse(response: AxiosResponse): void {
    if (response.status >= 300) {
      const msg = `❗ Something didn't go quite right. The API returned a non-ok status code ${response.status} with output: ${response.data}`;
      // TODO: Better error handling
      throw new Error(msg);
    }
  }

  public async makeRequest<T>(
    request_method: Method,
    endpoint: Routes,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data?: string | Record<string, any> | null,
    params?: Record<string, unknown>,
  ): Promise<T> {
    // Check to see if our token is expired before making a request
    // and refresh token if it's expired
    if (![Routes.login, Routes.apiKeyLogin].includes(endpoint) && this.token) {
      const payload = decode(this.token, { json: true });
      if (payload?.exp && payload.exp < Math.floor(Date.now() / 1000)) {
        this.token = await this.getToken();
      }
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let headers: Record<string, any> = {};

    if (this.token) {
      headers = this.getAuthHeader(this.token);
    }

    const config: AxiosRequestConfig = {
      method: request_method,
      url: `${this.apiUrl}/${endpoint
        .replace("{project_id}", this.projectId)
        .replace("{run_id}", this.runId)}`,
      params,
      headers,
      data,
    };

    const response = await axios.request<T>(config);

    this.validateResponse(response);

    return response.data;
  }
}
